File: Platform/Workspace/CSharp/Allors.Workspace/Meta/IComposite.cs
// <copyright file="IComposite.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>
// <summary>Defines the ObjectType type.</summary>

namespace Allors.Workspace.Meta
{
    using System.Collections.Generic;

    public interface IComposite : IObjectType
    {
        IEnumerable<IAssociationType> DatabaseAssociationTypes { get; }

        IEnumerable<IRoleType> DatabaseRoleTypes { get; }

        bool ExistDatabaseClass { get; }

        IEnumerable<IClass> DatabaseClasses { get; }

        bool ExistExclusiveDatabaseClass { get; }

        IClass ExclusiveDatabaseClass { get; }

        bool IsSynced { get; }

        // TODO: change to HasSuperType
        bool ExistSupertype(IInterface @interface);

        bool ExistAssociationType(IAssociationType association);

        bool ExistRoleType(IRoleType roleType);

        bool IsAssignableFrom(IComposite objectType);
    }
}

File: System/Workspace/Typescript/libs/workspace/src/lib/IComposite.ts
import { IAssociationType } from './IAssociationType';
import { IClass } from './IClass';
import { IInterface } from './IInterface';
import { IMethodType } from './IMethodType';
import { IObjectType } from './IObjectType';
import { IRoleType } from './IRoleType';

export interface IComposite extends IObjectType {
  IsSynced: boolean;

  DirectSupertypes: IInterface[];

  Supertypes: IInterface[];

  Classes: IClass[];

  AssociationTypes: IAssociationType[];

  RoleTypes: IRoleType[];

  WorkspaceRoleTypes: IRoleType[];

  DatabaseRoleTypes: IRoleType[];

  MethodTypes: IMethodType[];

  IsAssignableFrom(objectType: IComposite): bool;
}
